File: leetcode_ly/java/Binary Search/GuessNumberHigherOrLower.java
/* The guess API is defined in the parent class GuessGame.
   @param num, your guess
   @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
      int guess(int num); */

public class Solution extends GuessGame {
    public int guessNumber(int n) {
        int i = 1;
        int j = n;
        //下面判断加=可以直接定位到要找的数，不加则最后要找的要么是i，要么是j
        while(i <= j){
            int m = i + (j - i) / 2;
            if(guess(m) == 0) return m;
            else if(guess(m) == 1) i = m + 1;
            else j = m - 1;
        }
        return -1;
    }
}

File: leetcode_ly/c++/BinarySearch/GuessNumberHigherOrLower.cpp
// Forward declaration of guess API.
// @param num, your guess
// @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
int guess(int num);

class Solution {
public:
    int guessNumber(int n) {
        int low = 1;
        while(low <= n){
            int middle = low + (n - low) / 2;
            if(guess(middle) == 1) low = middle + 1;
            else if(guess(middle) == -1) n = middle - 1;
            else return middle;
        }
        return -1;
    }
};